edb/server/protocol/binary.pyx

```diff
@@ -499,6 +499,28 @@ cdef class EdgeConnection(frontend.FrontendConnection):
         else:
             return edgeql.NormalizedSource.from_string(text)
 
+    async def _heartbeat(self, stop_event):
+        """A heartbeat function that keeps the connection from going idle.
```

Comment:
Probably a better fix would be to just set `idle_session_timeout` to `0` while the compiler is running?  Possibly delayed (to, say, 1/2 of the timeout in the future) so that fast compiles don't trigger it.  Postgres allows transaction-local configuration so this should be safe even if restoration of the original timeout fails somehow.

Reply:
The problem with the possibly delayed approach is that if a bunch of the timeout has already elapsed when the query is sent, then we can still timeout while compiling. That's why my version does the first select right away, instead of delayed.

Part of my rationale for going with this version is that then, the common case is one extra query and not two.

We can do it by clearing idle_session_timeout, though

Reply:
> Part of my rationale for going with this version is that then, the common case is one extra query and not two.

It's only affecting queries in transaction blocks so even 2 extra quick queries might be OK.